### PR TITLE
ARROW-6110: [Java][Integration] Support LargeList Type and add integration test with C++

### PR DESCRIPTION
**Pull request by rymurr**
Add large list and ensure it works with Integration tests. As noted in the JIRA
ticket this is rather limited as the underlying vector doesn't support int64 addressing

The important downcasts to int32 have been noted for a follow up once vectors with
long addresses are supported

### CODE REVIEW

**Comment by emkornfield** [comment]
@BryanCutler would you have time to review?

**Comment by emkornfield** [comment]
@rymurr looks like this needs a rebase

**Comment by rymurr** [comment]
> @rymurr looks like this needs a rebase

Thanks for the reminder @emkornfield, done.

**Comment by BryanCutler** [comment]
I'm a little swamped right now, but I'll try to review sometime this week

File: java/vector/src/main/codegen/templates/UnionLargeListWriter.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  @Override
+  public StructWriter struct(String name) {
+    StructWriter structWriter = writer.struct(name);
+    return structWriter;
+  }
+
+  @Override
+  public void startList() {
+    vector.startNewValue(index);
+    writer.setPosition(checkedCastToInt(vector.getOffsetBuffer().getLong((index + 1) * OFFSET_WIDTH)));
```
Comment (liyafan82) [inline java/vector/src/main/codegen/templates/UnionLargeListWriter.java:161]: Here we need to cast the argument of getLong to long, to avoid integer overflow.

File: java/vector/src/main/codegen/templates/UnionLargeListWriter.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  @Override
+  public void endList() {
+    vector.getOffsetBuffer().setLong((index + 1) * OFFSET_WIDTH, writer.idx());
```
Comment (liyafan82) [inline java/vector/src/main/codegen/templates/UnionLargeListWriter.java:166]: Here we need to cast the first argument, too.

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 3]
```diff
@@ -73,6 +87,28 @@ public static void setBit(ArrowBuf validityBuffer, int index) {
...
   }
 
+  /**
+   * Set the bit at provided index to 1.
+   *
+   * @param validityBuffer validity buffer of the vector
+   * @param index index to be set
+   */
+  public static void setBit(ArrowBuf validityBuffer, long index) {
+    // it can be observed that some logic is duplicate of the logic in setValidityBit.
```
Comment (liyafan82) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:97]: I think we can replace the old implementation 
```
void setBit(ArrowBuf validityBuffer, int index)
```
with this one, as now ArrowBuf is based on 64-bit index.

File: java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java [hunk 4]
```diff
@@ -494,4 +505,46 @@ protected boolean compareFixedSizeListVectors(Range range) {
...
+
+      boolean isNull = leftVector.isNull(leftIndex);
+      if (isNull != rightVector.isNull(rightIndex)) {
+        return false;
+      }
+
+      int offsetWidth = LargeListVector.OFFSET_WIDTH;
+
+      if (!isNull) {
+        final long startIndexLeft = leftVector.getOffsetBuffer().getLong(leftIndex * offsetWidth);
```
Comment (liyafan82) [inline java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:529]: Here we need to cast getLong argument to long.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 5]
```diff
@@ -0,0 +1,991 @@
...
+      throw new IllegalArgumentException("Child vector already existed: " + addOrGetVector.getVector());
+    }
+
+    addOrGetVector.getVector().initializeChildrenFromFields(field.getChildren());
+  }
+
+  @Override
+  public void setInitialCapacity(int numRecords) {
+    validityAllocationSizeInBytes = getValidityBufferSizeFromCount(numRecords);
+    offsetAllocationSizeInBytes = (numRecords + 1) * OFFSET_WIDTH;
```
Comment (liyafan82) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:147]: We need to cast (numRecords + 1) to long. Otherwise, integer overflow may happen before promoting the result to long.

File: java/vector/src/main/codegen/templates/UnionLargeListWriter.java (hunk 1 above)
Comment (rymurr) [reply java/vector/src/main/codegen/templates/UnionLargeListWriter.java:161]: Hey @liyafan82 not sure what you mean. `getLong` returns `long` and `checkedCastToInt` takes a `long`. Sorry if I am missing something obvious

Comment (rymurr) [reply java/vector/src/main/codegen/templates/UnionLargeListWriter.java:161]: done

File: java/vector/src/main/codegen/templates/UnionLargeListWriter.java (hunk 2 above)
Comment (rymurr) [reply java/vector/src/main/codegen/templates/UnionLargeListWriter.java:166]: done

File: java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java (hunk 4 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:529]: done

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java (hunk 3 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:97]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java (hunk 5 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:147]: done

**Comment by nealrichardson** [comment]
Is this good to merge now? @BryanCutler are you still planning to review this? Would like to get this in 1.0.

**Comment by BryanCutler** [comment]
> Is this good to merge now? @BryanCutler are you still planning to review this? Would like to get this in 1.0.

I'm taking a look now, I'd like to get it in for 1.0 too.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java [hunk 6]
```diff
@@ -564,8 +564,10 @@ private ArrowBuf readIntoBuffer(BufferAllocator allocator, BufferType bufferType
     } else if (bufferType.equals(OFFSET)) {
       if (type == Types.MinorType.LARGEVARCHAR || type == Types.MinorType.LARGEVARBINARY) {
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java:565]: could you add a check for `Types.MinorType.LARGELIST` here are remove the changes below?

Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java:565]: you also won't need `typeBitWidth` as an arg

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Thanks for working on this @rymurr !  Apologies for taking so long to review.. It looks pretty good, but I saw what looked like inconsistencies in the `LargeListVector` APIs using ints vs longs to me, and otherwise only minor things to fix up.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 7]
```diff
@@ -0,0 +1,991 @@
...
+        offsetBuffer.setLong((i + 1) * OFFSET_WIDTH, currentOffset);
+      }
+    }
+    /* valueCount for the data vector is the current end offset */
+    final long childValueCount = (valueCount == 0) ? 0 :
+            offsetBuffer.getLong((lastSet + 1) * OFFSET_WIDTH);
+    /* set the value count of data vector and this will take care of
+     * checking whether data buffer needs to be reallocated.
+     */
+    vector.setValueCount(checkedCastToInt(childValueCount));
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:971]: Also indicate this in the javadoc here and probably for the class. It might even be a good idea to raise an error if the user tries to add too many elements, otherwise things might just start looking wrong.

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java (hunk 3 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:75]: done

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 8]
```diff
@@ -37,6 +37,20 @@
...
+   * Get the index of byte corresponding to bit index in validity buffer.
+   */
+  public static long byteIndex(long absoluteBitIndex) {
+    return absoluteBitIndex >> 3;
+  }
+
+  /**
+   * Get the relative index of bit within the byte in validity buffer.
+   */
+  public static long bitIndex(long absoluteBitIndex) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:50]: done

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 9]
```diff
@@ -57,12 +71,12 @@ public static int bitIndex(int absoluteBitIndex) {
...
    */
-  public static void setBit(ArrowBuf validityBuffer, int index) {
+  public static void setBit(ArrowBuf validityBuffer, long index) {
     // it can be observed that some logic is duplicate of the logic in setValidityBit.
     // this is because JIT cannot always remove the if branch in setValidityBit,
     // so we give a dedicated implementation for setting bits.
-    final int byteIndex = byteIndex(index);
-    final int bitIndex = bitIndex(index);
+    final long byteIndex = byteIndex(index);
+    final long bitIndex = bitIndex(index);
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:79]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 10]
```diff
@@ -0,0 +1,991 @@
...
+  private void allocateValidityBuffer(final long size) {
+    final int curSize = (int) size;
+    validityBuffer = allocator.buffer(curSize);
+    validityBuffer.readerIndex(0);
+    validityAllocationSizeInBytes = curSize;
+    validityBuffer.setZero(0, validityBuffer.capacity());
+  }
+  
+  protected void allocateOffsetBuffer(final long size) {
+    final int curSize = (int) size;
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:335]: yup! Done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 11]
```diff
@@ -0,0 +1,991 @@
...
+    reallocOffsetBuffer();
+    vector.reAlloc();
+  }
+
+  private void reallocValidityAndOffsetBuffers() {
+    reallocOffsetBuffer();
+    reallocValidityBuffer();
+  }
+  
+  protected void reallocOffsetBuffer() {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:360]: not sure how that happened. I have fixed it though.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 12]
```diff
@@ -0,0 +1,991 @@
...
+
+    final ArrowBuf newBuf = allocator.buffer(newAllocationSize);
+    newBuf.setBytes(0, offsetBuffer, 0, currentBufferCapacity);
+    newBuf.setZero(currentBufferCapacity, newBuf.capacity() - currentBufferCapacity);
+    offsetBuffer.getReferenceManager().release(1);
+    offsetBuffer = newBuf;
+    offsetAllocationSizeInBytes = newAllocationSize;
+  }
+  
+  private void reallocValidityBuffer() {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:385]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 13]
```diff
@@ -0,0 +1,991 @@
...
+    Preconditions.checkArgument(this.getMinorType() == from.getMinorType());
+    FieldReader in = from.getReader();
+    in.setPosition(inIndex);
+    UnionLargeListWriter out = getWriter();
+    out.setIndex(outIndex);
+    ComplexCopier.copy(in, out);
+  }
+
+  @Override
+  public UInt4Vector getOffsetVector() {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:440]: agreed, done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 14]
```diff
@@ -0,0 +1,991 @@
...
+  }
+
+  @Override
+  public Field getField() {
+    return new Field(getName(), fieldType, Collections.singletonList(getDataVector().getField()));
+  }
+
+  @Override
+  public MinorType getMinorType() {
+    return MinorType.LIST;
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:749]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 15]
```diff
@@ -0,0 +1,991 @@
...
+  }
+
+  @Override
+  public String getName() {
+    return null;
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:754]: yep, done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 16]
```diff
@@ -0,0 +1,991 @@
...
+    BitVectorHelper.setBit(validityBuffer, index);
+    lastSet = index;
+  }
+
+  /**
+   * Start a new value in the list vector.
+   *
+   * @param index index of the value to start
+   */
+  public long startNewValue(long index) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:922]: correct, fixed

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 17]
```diff
@@ -0,0 +1,991 @@
...
+    while (index >= getValidityAndOffsetValueCapacity()) {
+      reallocValidityAndOffsetBuffers();
+    }
+    for (long i = lastSet + 1; i <= index; i++) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:926]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 18]
```diff
@@ -0,0 +1,991 @@
...
+      final long currentOffset = offsetBuffer.getLong(i * OFFSET_WIDTH);
+      offsetBuffer.setLong((i + 1) * OFFSET_WIDTH, currentOffset);
+    }
+    BitVectorHelper.setBit(validityBuffer, index);
+    lastSet = index;
+    return checkedCastToInt(offsetBuffer.getLong((lastSet + 1) * OFFSET_WIDTH));
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:932]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 19]
```diff
@@ -0,0 +1,991 @@
...
+  }
+
+  /**
+   * End the current value.
+   *
+   * @param index index of the value to end
+   * @param size  number of elements in the list that was written
+   */
+  public void endValue(int index, int size) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:941]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 20]
```diff
@@ -0,0 +1,991 @@
...
+   */
+  @Override
+  public void setValueCount(int valueCount) {
+    this.valueCount = valueCount;
+    if (valueCount > 0) {
+      while (valueCount > getValidityAndOffsetValueCapacity()) {
+        /* check if validity and offset buffers need to be re-allocated */
+        reallocValidityAndOffsetBuffers();
+      }
+      for (long i = lastSet + 1; i < valueCount; i++) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:959]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 21]
```diff
@@ -0,0 +1,991 @@
...
+     * checking whether data buffer needs to be reallocated.
+     */
+    vector.setValueCount(checkedCastToInt(childValueCount));
+  }
+
+  public void setLastSet(int value) {
+    lastSet = value;
+  }
+
+  public long getLastSet() {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:978]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:971]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 22]
```diff
@@ -0,0 +1,991 @@
...
+    return lastSet;
+  }
+
+  @Override
+  public int getElementStartIndex(int index) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:983]: done

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 23]
```diff
@@ -0,0 +1,991 @@
...
+    return checkedCastToInt(offsetBuffer.getLong(index * OFFSET_WIDTH));
+  }
+
+  @Override
+  public int getElementEndIndex(int index) {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:988]: done

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java (hunk 6 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java:565]: thanks, done

File: java/vector/src/main/codegen/templates/UnionLargeListWriter.java [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+
+    import static org.apache.arrow.memory.util.LargeMemoryUtil.checkedCastToInt;
+<#include "/@includes/vv_imports.ftl" />
+
+/*
+ * This class is generated using freemarker and the ${.template_name} template.
+ */
+
+@SuppressWarnings("unused")
+public class UnionLargeListWriter extends AbstractFieldWriter {
```
[the remark this answers is not in the document]
Comment (rymurr) [reply java/vector/src/main/codegen/templates/UnionLargeListWriter.java:41]: agreed, I fixed this...was just being lazy ;-)

**Comment by rymurr** [comment]
> Thanks for working on this @rymurr ! Apologies for taking so long to review.. It looks pretty good, but I saw what looked like inconsistencies in the `LargeListVector` APIs using ints vs longs to me, and otherwise only minor things to fix up.

Thanks a lot for the thorough review. I have fixed up everything you mentioned. It appears some of the confusion was related to changes for 64-bit allocations that were recently merged and the rest was my ignorance!

I have pushed a change with all your recommended fixes.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 25]
```diff
@@ -0,0 +1,1004 @@
...
+   * @param index position of the element
+   * @return Object at given position
+   */
+  @Override
+  public Object getObject(int index) {
+    if (isSet(index) == 0) {
+      return null;
+    }
+    final List<Object> vals = new JsonStringArrayList<>();
+    final long start = offsetBuffer.getLong(index * OFFSET_WIDTH);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:830]: does this arg need to be cast to long?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 26]
```diff
@@ -0,0 +1,1004 @@
...
+    final long end = offsetBuffer.getLong((index + 1) * OFFSET_WIDTH);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:831]: also here

File: java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java [hunk 27]
```diff
@@ -0,0 +1,1004 @@
...
+  /**
+   * Check if element at given index is empty list.
+   * @param index position of element
+   * @return true if element at given index is empty list or NULL, false otherwise
+   */
+  public boolean isEmpty(int index) {
+    if (isNull(index)) {
+      return true;
+    } else {
+      final long start = offsetBuffer.getLong(index * OFFSET_WIDTH);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java:860]: cast to long?

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java [hunk 28]
```diff
@@ -710,7 +712,10 @@ private void readFromJsonIntoVector(Field field, FieldVector vector) throws Json
           innerBufferValueCount = valueCount + 1;
         }
 
-        vectorBuffers[v] = readIntoBuffer(allocator, bufferType, vector.getMinorType(), innerBufferValueCount);
+        vectorBuffers[v] = readIntoBuffer(allocator,
+            bufferType,
+            vector.getMinorType(),
+            innerBufferValueCount);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java:718]: was there a reason for this change?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/impl/UnionLargeListReader.java [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+  public void setPosition(int index) {
+    setIndex(index);
+  }
+
+  /**
+   * set current index.
+   *
+   * This is different from setPosition in that it supports int64
+   */
+  public void setIndex(long index) {
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/impl/UnionLargeListReader.java:68]: why is this needed? the vector only supports integer max_value number of elements right?

File: java/vector/src/main/java/org/apache/arrow/vector/validate/ValidateVectorVisitor.java [hunk 30]
```diff
@@ -135,6 +136,45 @@ public Void visit(ListVector vector, Void value) {
...
+        throw new IllegalArgumentException("valueBuffer is null or capacity is 0");
+      }
+
+      if (dataExtent > dataVector.getValueCount()) {
+        throw new IllegalArgumentException(String.format("Length spanned by list offsets (%s) larger than" +
+            " data vector valueCount (length %s)", dataExtent, dataVector.getValueCount()));
+      }
+    }
+
+    return dataVector.accept(this, null);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/validate/ValidateVectorVisitor.java:175]: doesn't `value` need to be passed in here?

**Review by BryanCutler — COMMENTED** [review]
Thanks for the quick update @rymurr , it looks pretty good! Only a couple minor things. I see quite a few instances of `offsetBuffer.getLong/setLong(i * OFFSET_WIDTH)` that I believe need to be cast to long to avoid overflow. Could you take a quick pass and fix those up? I think we will be good to go after that.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java (hunk 28 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/JsonFileReader.java:718]: IDE autoformatting...reverted!

File: java/vector/src/main/java/org/apache/arrow/vector/complex/impl/UnionLargeListReader.java (hunk 29 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/impl/UnionLargeListReader.java:68]: you are correct. Removed

File: java/vector/src/main/java/org/apache/arrow/vector/validate/ValidateVectorVisitor.java (hunk 30 above)
Comment (rymurr) [reply java/vector/src/main/java/org/apache/arrow/vector/validate/ValidateVectorVisitor.java:175]: value is always of type `Void`. None of the other visitors pass value to children.

**Comment by rymurr** [comment]
> Thanks for the quick update @rymurr , it looks pretty good! Only a couple minor things. I see quite a few instances of `offsetBuffer.getLong/setLong(i * OFFSET_WIDTH)` that I believe need to be cast to long to avoid overflow. Could you take a quick pass and fix those up? I think we will be good to go after that.

Thanks for another thorough review @BryanCutler ! I think that I addressed everything. I may have went over the top w/ the casting but it doesn't hurt :-)

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @rymurr !
